### ISSUE TITLE
Impossible to change run ID if wandb.init() is called after wandb.agent()

### ISSUE THREAD

**Issue by tommasocarraro** [wandb/wandb#4653]
### Describe the bug

Suppose one wants to perform a hyper-parameter step in an ML pipeline and then wants to learn a model with the best hyper-parameter configuration found. Suppose also that one wants to perform this two steps within the same process. 
In Weights and Biases, this can be done by performing a sweep, with a call to wandb.agent(), followed by a run that trains the model with the best hyper-parameters found by the wandb agent. This second step can be obtained by calling wandb.init().

However, if you call wandb.init() after a wandb.agent() call, the run created by wandb.init() is not a new run, as one expects. Instead, it is exactly the last run of the sweep, again (?). In other words, the last run of the sweep seems to be still alive when I call wandb.init(). So, even if I call wandb.init() to create a new run, wandb is still referring to the last sweep run. I am not able to understand how this is possible. Even if I force wandb to change the ID of the run explicitly, the run ID does not change.

I found a similar issue https://github.com/wandb/wandb/issues/1529, however, my case is different since the issue argues this problem within the runs in the same sweep. In my case, we have this problem between the last run of the sweep and the first run created after the sweep. I tried the solution proposed in the issue but, as you can see in the code snippet and in the log information, it does not seem to work.

Here, I post a simple code snippet to show the problem. I simply defined a trivial sweep, with no-sense content on it, and then called a wandb.agent() followed by a wandb.init(). Here, we have to pay attention to the printed IDs of the runs and their names. As you can see, the first two runs of the sweep are just ok, with different IDs and names. Then, the third run seems to be still alive when I call wandb.init() after the sweep is ended. In fact, even if I set reinit=True and force the run to have a new ID, the ID of the last run of the sweep is preserved on the new run I am trying to create. In other words, they are the same run. However, it seems it is possible to change its name, as you can see from the logged information (I highlighted IDs and names of runs in bold in order to find them in the log). In addition, you can see that the log shows that the run is detected as a sweep run, even if it is out of the context of the agent and run after the agent should have finished its task. 

```python
import wandb
import random
import os


def train():
    with wandb.init(project="ciao") as run:
        print(run.id)
        run.name = "run_1"
        print(run.name)
        for i in range(10):
            wandb.log({"metric": i * random.randint(0, 10)})


def reset_wandb_env():
    exclude = {
        "WANDB_PROJECT",
        "WANDB_ENTITY",
        "WANDB_API_KEY",
    }
    for k, v in os.environ.items():
        if k.startswith("WANDB_") and k not in exclude:
            del os.environ[k]


if __name__ == "__main__":
    sweep_id = wandb.sweep({
      "method": "random",
      "metric": {"goal": "maximize", "name": "metric"},
      "parameters":
          {
            "p1": {"values": [1, 2, 3, 4]},
            "p2": {"values": [9, 3]}}}, project="ciao")
    # find the best model with a sweep
    wandb.agent(sweep_id, function=train, count=3)
    reset_wandb_env()
    with wandb.init(project="ciao", reinit=True, id="new_id", name="run_2") as run:
        # training of the best model with a simple run
        print(run.id)
        print(run.name)
```

### Additional Files

Create sweep with ID: shfz89oo
Sweep URL: https://wandb.ai/bmxitalia/ciao/sweeps/shfz89oo
wandb: Agent Starting Run: vnfoctm9 with config:
wandb: 	p1: 3
wandb: 	p2: 9
wandb: Currently logged in as: bmxitalia. Use `wandb login --relogin` to force relogin
wandb: WARNING Ignored wandb.init() arg project when running a sweep.
wandb: wandb version 0.13.7 is available!  To upgrade, please run:
wandb:  $ pip install wandb --upgrade
wandb: Tracking run with wandb version 0.13.6
wandb: Run data is saved locally in /Users/ciomi/PycharmProjects/LTNrec-knowledge-transfer/wandb/run-20221216_200341-vnfoctm9
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run robust-sweep-1
wandb: ⭐️ View project at https://wandb.ai/bmxitalia/ciao
wandb: 🧹 View sweep at https://wandb.ai/bmxitalia/ciao/sweeps/shfz89oo
wandb: 🚀 View run at https://wandb.ai/bmxitalia/ciao/runs/vnfoctm9
wandb: Waiting for W&B process to finish... (success).
**vnfoctm9
robust-sweep-1**
wandb: 
wandb: Run history:
wandb: metric ▁▂▂▃▃▄▆▂█▆
wandb: 
wandb: Run summary:
wandb: metric 54
wandb: 
wandb: Synced robust-sweep-1: https://wandb.ai/bmxitalia/ciao/runs/vnfoctm9
wandb: Synced 5 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)
wandb: Find logs at: ./wandb/run-20221216_200341-vnfoctm9/logs
wandb: Agent Starting Run: 6914adze with config:
wandb: 	p1: 2
wandb: 	p2: 3
wandb: WARNING Ignored wandb.init() arg project when running a sweep.
wandb: wandb version 0.13.7 is available!  To upgrade, please run:
wandb:  $ pip install wandb --upgrade
wandb: Tracking run with wandb version 0.13.6
wandb: Run data is saved locally in /Users/ciomi/PycharmProjects/LTNrec-knowledge-transfer/wandb/run-20221216_200351-6914adze
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run summer-sweep-2
wandb: ⭐️ View project at https://wandb.ai/bmxitalia/ciao
wandb: 🧹 View sweep at https://wandb.ai/bmxitalia/ciao/sweeps/shfz89oo
wandb: 🚀 View run at https://wandb.ai/bmxitalia/ciao/runs/6914adze
wandb: Waiting for W&B process to finish... (success).
**6914adze
summer-sweep-2**
wandb: 
wandb: Run history:
wandb: metric ▁▂▂▃▁▇▅█▄█
wandb: 
wandb: Run summary:
wandb: metric 36
wandb: 
wandb: Synced summer-sweep-2: https://wandb.ai/bmxitalia/ciao/runs/6914adze
wandb: Synced 5 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)
wandb: Find logs at: ./wandb/run-20221216_200351-6914adze/logs
wandb: Agent Starting Run: 17grafsg with config:
wandb: 	p1: 3
wandb: 	p2: 9
wandb: WARNING Ignored wandb.init() arg project when running a sweep.
wandb: wandb version 0.13.7 is available!  To upgrade, please run:
wandb:  $ pip install wandb --upgrade
wandb: Tracking run with wandb version 0.13.6
wandb: Run data is saved locally in /Users/ciomi/PycharmProjects/LTNrec-knowledge-transfer/wandb/run-20221216_200403-17grafsg
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run fluent-sweep-3
wandb: ⭐️ View project at https://wandb.ai/bmxitalia/ciao
wandb: 🧹 View sweep at https://wandb.ai/bmxitalia/ciao/sweeps/shfz89oo
wandb: 🚀 View run at https://wandb.ai/bmxitalia/ciao/runs/17grafsg
wandb: Waiting for W&B process to finish... (success).
**17grafsg
fluent-sweep-3**
wandb: 
wandb: Run history:
wandb: metric ▁▂▂▁▂▅▁▂▅█
wandb: 
wandb: Run summary:
wandb: metric 90
wandb: 
wandb: Synced fluent-sweep-3: https://wandb.ai/bmxitalia/ciao/runs/17grafsg
wandb: Synced 5 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)
wandb: Find logs at: ./wandb/run-20221216_200403-17grafsg/logs
wandb: WARNING Changes to your `wandb` environment variables will be ignored because your `wandb` session has already started. For more information on how to modify your settings with `wandb.init()` arguments, please refer to https://wandb.me/wandb-init.
wandb: WARNING Ignored wandb.init() arg project when running a sweep.
wandb: WARNING Ignored wandb.init() arg id when running a sweep.
wandb: wandb version 0.13.7 is available!  To upgrade, please run:
wandb:  $ pip install wandb --upgrade
wandb: Tracking run with wandb version 0.13.6
wandb: Run data is saved locally in /Users/ciomi/PycharmProjects/LTNrec-knowledge-transfer/wandb/run-20221216_200408-17grafsg
wandb: Run `wandb offline` to turn off syncing.
wandb: Syncing run run_1
wandb: ⭐️ View project at https://wandb.ai/bmxitalia/ciao
wandb: 🧹 View sweep at https://wandb.ai/bmxitalia/ciao/sweeps/shfz89oo
wandb: 🚀 View run at https://wandb.ai/bmxitalia/ciao/runs/17grafsg
wandb: Waiting for W&B process to finish... (success).
**17grafsg
run_1**
wandb: Synced run_1: https://wandb.ai/bmxitalia/ciao/runs/17grafsg
wandb: Synced 5 W&B file(s), 0 media file(s), 0 artifact file(s) and 0 other file(s)
wandb: Find logs at: ./wandb/run-20221216_200408-17grafsg/logs

### Environment

WandB version: 0.13.6

OS: macos 12.5

Python version: 3.9

**Comment by tommasocarraro** [comment]
I want to add that I could have understood what is the cause of this problem.

Simply, once a sweep is started, it remains forever in the RUNNING status. I checked this on the UI.

I made this check for this toy project and also for a project that I run more than 1 year ago. A sweep from one year ago is still in the RUNNING status.

Is it possible that this is the problem? How is it possible to manually change the status after the call to wandb.agent()?

**Comment by tommasocarraro** [comment]
I have discovered another important piece of information.

The final and real problem is that sweeps with random strategy never finish, even if the `count` parameter of `wandb.agent()` is defined.

For this reason, the user cannot create new runs after `wandb.agent()` because the agent will run forever and the `RUN_ID` will be always the same for all the runs. I have tried with grid search, and since wandb sets the sweep to the FINISHED status at the end of the predefined runs, it is then possible to call `wandb.init()` without any problems.

So, the issue is still open and the user needs a way to stop the sweep using a command on the python script. Or, at least, it will be appreciated an improvement that marks the sweep as finished when the number of runs matches the predefined number of runs defined with the `count` parameter. This should be easy to implement as it should be very similar to what it has been done for the `grid` strategy. For the moment, it is only possible to stop the sweep from the UI.

**Comment by ramit-wandb** [comment]
Hi @tommasocarraro,

Thanks for reporting this - this is already a known issue which is on the roadmap to be resolved, but hasn't been worked on as of now. I would love to understand your use case to run a normal run in the same script as you are using for hyperparameter optimization.

I'll be reaching out to our engineering teams to resolve this issue on our end. 

Thanks,
Ramit

**Comment by tommasocarraro** [comment]
Hi @ramit-wandb, thank you very much for your reply!

The use case is really easy. I would like to create an artifact containing a JSON file with the best hyper-parameter values found during the random search. 

For doing that, we need to wait for the sweep to end. Then, we call the API to take the best run, and its corresponding hyper-parameter values. Finally, we can take the dictionary returned and create a JSON file.

After all these steps, in order to create an artifact, we need to start a new run to log the artifact on the server. However, it is not possible to do that because this last run will overwrite the last sweep run since sweeps for random and bayesian searches never end. One does not want to overwrite and lose all the precious data contained in the sweep run.

Let me know if you need further information.

**Comment by ramit-wandb** [comment]
Got it, thanks for the explanation! I'm going to increase the priority on our internal ticket for this, and we will reach out once this issue has been resolved.

**Comment by Gumichocopengin8** [comment]
@ramit-wandb any update on this? I got the same issue with `wandb==0.16.2`.

**Comment by ijdoc** [comment]
@tommasocarraro @Gumichocopengin8 please try running the following just before your `wandb.init()` call:
```python
wandb.teardown()
```
please do add a comment to indicate if this solves the issue
